contrib/src/main/java/com/datatorrent/contrib/hbase/HBasePOJOInputOperator.java

```diff
@@ -69,116 +69,103 @@ public Object convertValue( HBaseFieldInfo fieldInfo, Object value)
-  }
-
-  @Override
-  public void emitTuples()
-  {
-    try {
-      Scan scan = nextScan();
-      if (scan == null)
-        return;
-
-      ResultScanner resultScanner = store.getTable().getScanner(scan);
-
-      while (true) {
-        Result result = resultScanner.next();
-        if (result == null)
-          break;
-
-        String readRow = Bytes.toString(result.getRow());
-        if( readRow.equals( lastReadRow ))
-          continue;
-
-        Object instance = pojoType.newInstance();
-        rowSetter.set(instance, readRow);
-
-        List<Cell> cells = result.listCells();
+      String readRow = Bytes.toString(result.getRow());
+      if( readRow.equals( getLastReadRow() )) {
+        return null;
+      }
 
-        for (Cell cell : cells) {
-          String columnName = Bytes.toString(CellUtil.cloneQualifier(cell));
-          byte[] value = CellUtil.cloneValue(cell);
-          fieldValueGenerator.setColumnValue( instance, columnName, value, valueConverter );
-        }
+      Object instance = pojoType.newInstance();
+      rowSetter.set(instance, readRow);
 
-        outputPort.emit(instance);
-        lastReadRow = readRow;
+       List<Cell> cells = result.listCells();
+       for (Cell cell : cells) {
+         String columnName = Bytes.toString(CellUtil.cloneQualifier(cell));
+         String columnFamily = Bytes.toString(CellUtil.cloneFamily(cell));
+        byte[] value = CellUtil.cloneValue(cell);
+         ((HBaseFieldValueGenerator)fieldValueGenerator).setColumnValue(instance, columnName, columnFamily, value,
+             valueConverter);
       }
 
+      setLastReadRow(readRow);
+      return instance;
     } catch (Exception e) {
-      throw new RuntimeException(e.getMessage());
+      throw new RuntimeException(e);
     }
-
-  }
-
-  protected Scan nextScan()
-  {
-    if(lastReadRow==null && startRow==null )
-      return new Scan();
-    else
-      return new Scan( Bytes.toBytes( lastReadRow == null ? startRow : lastReadRow ) );
   }
 
-  public HBaseStore getStore()
-  {
-    return store;
-  }
-  public void setStore(HBaseStore store)
+  @Override
+  protected Scan operationScan()
   {
-    this.store = store;
+    Scan scan;
+    if (getLastReadRow() == null && getStartRow() == null) {
+      // If no start row specified and no row read yet
+      scan = new Scan();
+    } else if (getEndRow() == null) {
+      // If only start row specified
+      scan = new Scan(Bytes.toBytes(getLastReadRow() == null ? getStartRow() : getLastReadRow()));
+    } else {
+      // If end row also specified
+      scan = new Scan(Bytes.toBytes(getLastReadRow() == null ? getStartRow() : getLastReadRow()),
+          Bytes.toBytes(getEndRow()));
+    }
+    for (HBaseFieldInfo field : tableInfo.getFieldsInfo()) {
+      scan.addColumn(Bytes.toBytes(field.getFamilyName()), Bytes.toBytes(field.getColumnName()));
+    }
+    scan.setAttribute(Scan.HINT_LOOKAHEAD, Bytes.toBytes(getHintScanLookahead()));
+    return scan;
```

Comment:
Can you reuse the scan variable? That will save connection time for every request.